4단계 - 로또(2등)

haneepark opened the pull request:
4단계 기능 구현, 그리고 3단계에서 의견 주신 내용 추가로 반영했습니다. 

그리고..
* '코드리뷰 의견 반영'이라는 각 커밋 메시지의 Description에 어떤 내용을 반영했는지 간략하게 적어두었습니다.
* 추가 질문 하나 남겼는데, 댓글 확인 부탁드립니다: https://github.com/next-step/java-lotto-pro/pull/205#discussion_r745590848

예리하고 꼼꼼한 리뷰 좋았어요. 이번 리뷰도 미리 감사합니다 :)

haneepark committed:
docs(README): 구현할 기능 목록 작성

haneepark committed:
refactor: 클래스 이름 변경(LottoNumbers > LottoTicket)

haneepark committed:
feat(WinTicket): 보너스 번호를 가지는 로또 당첨 번호 클래스 구현

haneepark committed:
refactor: 코드리뷰 의견 반영

- 메소드명 변경`getNumberOfLottosPurchasableWith` > `countPurchasable`
- 변수명 변경 `list` > `lottoNumbers`
- `Lottos.getSellingPrice` stream 사용하지 않고 public 상수 쓰도록
- 메소드명 변경 `LottoNumbers.countIfMatch`> `count`

haneepark committed:
refactor: 코드리뷰 의견 반영

 - NumberListGenerator의 메소드를 Lottos 클래스로 이동

haneepark committed:
refactor(LottoConstants): 로또 관련 상수들을 한 클래스에서 관리하도록 수정

haneepark committed:
refactor(LottoTicket): 코드리뷰 의견 반영

- LottoTicket과 WinTicket의 생성자를 protected로 변경, 테스트 하지 않도록 해서 테스트 코드 중 중복 코드(generateLottoNumberListOf)를 제거. 대신 인스턴스 생성을 위한 public method인 `of`를 테스트 하도록 함.

haneepark committed:
feat(Lotto, ResultView): 로또 당첨 금액을 구할 때 보너스 번호가 일치하는 경우에 대한 상금을 추가하는 기능 구현

haneepark committed:
refactor(Money): 코드 리뷰 의견 반영

- 의미 없는 변수명 변경

kang-hyungu commented on src/main/java/lotto/LottoConstants.java:
```diff
@@ -0,0 +1,10 @@
+
+import lotto.model.Money;
+
+public class LottoConstants {
```
이 클래스 제거해주세요.
1. 전역변수랑 다를게 없습니다.
2. 객체지향적이지 않습니다. 책임을 한 곳으로 모으지 마세요. 적절한 객체로 각 상수를 옮겨주세요.

haneepark replied:
반영했습니다 :)
* 86155c0

kang-hyungu commented on src/main/java/lotto/LottoController.java:
```diff
@@ -28,22 +28,24 @@ private int getBuyMoney() {
     }
 
-    private LottoNumbers getWinLottoNumbers() {
+    private WinTicket getWinLottoTicket() {
```
이 메서드는 InputView에서 처리하는게 적절하지 않을까요?
어떤 메시지를 출력하고 어떤 값을 받아올지 입력 순서는 InputView에게 역할을 맡기신거 아닐가요?

haneepark replied:
맞네요. InputView에서 해당 메서드를 처리하도록 수정했습니다.
* 94e684c

kang-hyungu commented on src/main/java/lotto/LottoController.java:
```diff
@@ -28,22 +28,24 @@ private int getBuyMoney() {
-        final LottoResult lottoResult = lottos.calculateWinning(winNumbers);
-        final double roi = lottoResult.getReturnOnInvestment(lottos.getSellingPrice());
-        resultView.showWinningStatistics(lottoResult);
+    private void printResult(Lottos lottos, WinTicket winTicket) {
```
이 메서드는 ResultView가 적절해보이네요

haneepark replied:
ResultView에서 해당 메서드를 처리하도록 수정했습니다.
* 94e684c

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -1,43 +1,39 @@
+
 import java.util.List;
 
 public class Lotto {
```
Lotto 객체와 LottoTicket 객체는 어떤 차이가 있나요?
하나로 묶을 수 있지 않나요? 각각 어떤 역할이 있는 건가요?

haneepark replied:
맞네요. 계속 저도 뭔가 이상하게 느껴졌는데 이것 때문이었네요! 감사합니다. 반영했습니다! 
* 2b92ae7

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -1,43 +1,39 @@
-
-    public Money getSellingPrice() {
-        return SELLING_PRICE;
+    public Winning calculateWinning(WinTicket winTicket) {
```
이렇게 바꾸면 어떨까요?
값을 받아와서 당첨을 찾는게 아닌 필요한 값을 전달해서 당첨을 찾아달라는 메시지를 보내도록 수정해보아요

```java
    public Winning calculateWinning(WinTicket winTicket) {
        return ticket.getWinning(winTicket); // getWinning 메서드는 더 적절한 이름으로 바꿔주세요.
    }
```

haneepark replied:
좀 더 객체지향스럽게 수정했습니다.

* 040f04e

```java
public Winning calculateWinning(WinTicket winTicket) {
        final int count = winTicket.calculateNumberOfMatch(this);
        final boolean bonus = winTicket.matchBonusNumber(numbers);
        return Winning.of(count, bonus);
    }
```

kang-hyungu commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -36,17 +37,31 @@ public int size() {
         return new LottoResult(winnings);
     }
 
     public Money getSellingPrice() {
```
메서드명에 get을 쓰는게 적절할까요?

haneepark replied:
적절하지 않은가요? 단순히 반환이 아니라 계산을 해서 반환하기 때문에 get을 쓰면 이상하다는 말씀일까요? 제가 보기엔 괜찮은것같은데요. get하고 상관없이 `getTotalSellingPrice`라고 더 정확히 네이밍해도 괜찮겠다 싶기는 합니다만..

질문하신 의도를 잘 모르겠는데 설명 해 주실 수 있을까요?

kang-hyungu replied:
이 메서드는 총 판매 금액을 구하는거 아닌가요?
SELLING_PRICE는 1000원으로 되어 있는데 getSellingPrice 메서드는 로또가격 * 구매갯수를 수행하고 있죠.
SELLING_PRICE의 getter가 아닌데 get이 붙어있어서 남긴 질문이었습니다.

kang-hyungu commented on src/main/java/lotto/model/WinTicket.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class WinTicket extends LottoTicket {
```
상속은 제거해주세요.
자식 클래스들이 부모 클래스에 강하게 엮이게 되면서 변경에 취약하게 됩니다.
이펙티브 자바 3판 아이템 18. 상속보다는 컴포지션을 사용하라 
또는
조영호님의 오브젝트 책 참고하세요

kang-hyungu replied:
Winning, WinTicket
일관된 명칭을 써주세요

haneepark replied:
상속 제거하고 컴포지션 사용하도록 리팩토링했습니다.
* 040f04e

명칭은 아래와 같은 의미로 일관되게 사용했는데, 이상한 부분이 있나요? 아니면 이름을 잘못지은것일까요? 그렇다면 더 적절한 이름 추천해주실수있을까요?
* `Winning`: 각 로또 결과(등수)
* `WinTicket`: 당첨번호

kang-hyungu replied:
1. Winning으로 작성하셔서 WinningTickect을 써야 되는거 아닌가라는 질문이었는데요
2. WinTicket은 당첨티켓으로 읽힙니다.
당첨번호를 의도하신거라면 이름 조정이 필요해보입니다.
그리고 위에 Winning은 보통 당첨으로 이해하지 않을까요?
클래스명으로 등수, 로또 결과를 처리하는 객체로 인식하기 어렵진 않을까요?

kang-hyungu reviewed (changes requested):
하늬님 진행해주신 미션에 피드백 남겼습니다.
질문 주신 내용도 코멘트 남겼습니다.
참고하셔서 수정해보시고 궁금한 점 생기면 dm이나 코멘트 남겨주세요.

haneepark committed:
refactor(LottoTicket, Winning): 코드 리뷰 의견 반영

- Lotto 클래스 내용을 LottoTicket 클래스에 병합
- 결과에 맞는 Winning 객체를 반환하는 메소드 리팩토링 (Winning.ofMatchCount -> Winning.of)

haneepark committed:
refactor(LottoConstants): 로또 관련 상수들을 각 담당 클래스에서 관리하도록 수정

haneepark committed:
refactor(LottoGame): 클래스명 변경, View 리팩토링

- 클래스명 변경 (LottoController > LottoGame)
- LottoGame의 로직 일부를 InputView, ResultView로 이동

haneepark committed:
refactor(WinTicket): WinTicket의 상속 제거, LottoTicket.calculateWinning 로직을 객체지향스럽게 개선

kang-hyungu commented on src/main/java/lotto/model/Money.java:
```diff
@@ -30,8 +30,8 @@ public int hashCode() {
+    public Money multiplyBy(int multiplier) {
+        return new Money(this.money * multiplier);
     }
 
```
아래와 같은 코드가 동작하네요.
유효한 값일까요?

```java
    @Test
    void negative() {
        final Money money = new Money(-1000);
        assertThat(money.isZero()).isFalse();
    }
```

kang-hyungu commented on src/main/java/lotto/LottoGame.java:
```diff
@@ -0,0 +1,29 @@
+import lotto.view.InputView;
+import lotto.view.ResultView;
+
+public class LottoGame {
```
로또 게임이 엄청 깔끔해졌네요 👍

kang-hyungu commented on src/main/java/lotto/model/LottoTicket.java:
```diff
@@ -0,0 +1,60 @@
+    public static final int LOTTO_SIZE = 6;
+    private final List<LottoNumber> numbers;
+
+    protected LottoTicket(List<LottoNumber> numbers) {
```
```suggestion
    private LottoTicket(List<LottoNumber> numbers) {
```
상속을 없앴으니 접근제어자도 바꿔주는게 좋겠네요

kang-hyungu commented on src/main/java/lotto/model/WinTicket.java:
```diff
@@ -0,0 +1,27 @@
+
+    private final LottoNumber bonusNumber;
+
+    public WinTicket(LottoTicket lottoTicket, LottoNumber bonusNumber) {
```
null 체크도 추가해주면 어떨까요

kang-hyungu reviewed (approved):
하늬님 피드백 반영 잘 해주셨습니다.
추가로 남긴 피드백은 다음 미션에서 적용해주세요
pr은 머지 처리하겠습니다. 수고하셨습니다
